extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/internal/tsr/AgroalOrderedLastSynchronizationList.java

```diff
@@ -0,0 +1,156 @@
+                    sync.afterCompletion(status);
+                } catch (Exception e) {
+                    if (LOGGER.isDebugEnabled()) {
+                        LOGGER.debugf(
+                                "The synchronization %s associated with tx key %s failed during afterCompletion(%d): %s",
+                                sync, tsr.getTransactionKey(), status, e.getMessage());
+                    }
+                }
+            }
+        }
+    }
+
+    private final List<SynchronizationGroup> synchGroups = new ArrayList<>();
+    private SynchronizationGroup otherSyncs;
+    private final TransactionSynchronizationRegistry tsr;
+
+    public void setSynchronizationGroupOrder(String... packagePrefixes) {
+        for (var packagePrefix : packagePrefixes) {
+            var synchronizationGroup = new SynchronizationGroup(packagePrefix);
+
+            synchGroups.add(synchronizationGroup);
+
+            if (packagePrefix.isEmpty()) {
+                otherSyncs = synchronizationGroup;
+            }
+        }
+    }
+
+    public AgroalOrderedLastSynchronizationList(
+            TransactionSynchronizationRegistryWrapper transactionSynchronizationRegistryWrapper) {
+        this.tsr = transactionSynchronizationRegistryWrapper;
+
+        // order the synchronization groups as follows [ARC, OTHER, HIBERNATE, AGROAL].
+        // Agroal is last since it validates that connection wrappers are closed at the right time:
+        setSynchronizationGroupOrder("io.quarkus.arc.impl", "", "org.hibernate", "io.agroal.narayana");
+    }
+
+    /**
+     * This is only allowed at various points of the transaction lifecycle.
+     *
+     * @param synchronization The synchronization to register
+     * @throws IllegalStateException In case the transaction was in a state that was not valid to register under
+     * @throws SystemException In case the transaction status was not known
+     */
+    public void registerInterposedSynchronization(Synchronization synchronization)
+            throws IllegalStateException, SystemException {
+        int status = tsr.getTransactionStatus();
+
+        switch (status) {
+            case Status.STATUS_ACTIVE:
+            case Status.STATUS_PREPARING:
+                break;
+            case Status.STATUS_MARKED_ROLLBACK:
+                // do nothing; we can pretend like it was registered, but it'll never be run anyway.
+                return;
+            default:
+                throw new IllegalStateException("Syncs are not allowed to be registered when the tx is in state " + status);
+        }
+
+        // add the synchronization to the group that matches this package, if there is no matching group then add it to the
+        // catch-all group (otherSyncs)
+        String packageName = synchronization.getClass().getName();
+        SynchronizationGroup synchGroup = synchGroups.stream()
+                .filter(g -> packageName.startsWith(g.packageName)).findFirst().orElse(otherSyncs);
```

Comment:
Since this is the hot path (executed for every single request involving the database), would you mind using a simple loop instead of streams/lambdas here? I seem to recall capturing lambdas in particular have a bad impact on performance.

cc @Sanne / @gsmet

Reply:
Done